src/main/java/com/intuit/graphql/orchestrator/schema/transform/KeyTransformer.java

```diff
@@ -0,0 +1,115 @@
+package com.intuit.graphql.orchestrator.schema.transform;
+
+import static com.intuit.graphql.orchestrator.utils.FederationUtils.FEDERATION_KEY_DIRECTIVE;
+import static com.intuit.graphql.orchestrator.utils.FederationUtils.FEDERATION_REQUIRES_DIRECTIVE;
+import static com.intuit.graphql.orchestrator.utils.FederationUtils.containsExternalDirective;
+import static com.intuit.graphql.orchestrator.utils.FederationUtils.isBaseType;
+import static com.intuit.graphql.orchestrator.utils.XtextTypeUtils.getDirectivesFromDefinition;
+import static com.intuit.graphql.orchestrator.utils.XtextTypeUtils.getFieldDefinitions;
+import static com.intuit.graphql.orchestrator.utils.XtextUtils.typeContainsDirective;
+
+import com.intuit.graphql.graphQL.Argument;
+import com.intuit.graphql.graphQL.TypeDefinition;
+import com.intuit.graphql.graphQL.ValueWithVariable;
+import com.intuit.graphql.orchestrator.federation.keydirective.KeyDirectiveValidator;
+import com.intuit.graphql.orchestrator.federation.metadata.FederationMetadata;
+import com.intuit.graphql.orchestrator.federation.metadata.FederationMetadata.EntityExtensionMetadata;
+import com.intuit.graphql.orchestrator.federation.metadata.FederationMetadata.EntityMetadata;
+import com.intuit.graphql.orchestrator.federation.metadata.KeyDirectiveMetadata;
+import com.intuit.graphql.orchestrator.xtext.XtextGraph;
+import graphql.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.xtext.util.Strings;
+
+/**
+ * This class is responsible for checking the merged graph for any key directives. For each field in key
+ *  directive, this class will validate the fields to the key directive by ensuring they exist
+ */
+public class KeyTransformer implements Transformer<XtextGraph, XtextGraph> {
+
+  @VisibleForTesting
+  KeyDirectiveValidator keyDirectiveValidator = new KeyDirectiveValidator();
+
+  @Override
+  public XtextGraph transform(final XtextGraph source) {
+    if(source.getServiceProvider().isFederationProvider()) {
+      Map<String, TypeDefinition> entitiesByTypename = new HashMap<>();
+      Map<String, TypeDefinition> entityExtensionsByTypename = new HashMap<>();
```

Comment:
explore if entitiesByTypename  and entityExtensionsByTypename  can be merged into FederationData